apisix/discovery/nacos/init.lua

```diff
@@ -277,69 +288,118 @@ local function is_grpc(scheme)
-                weight = host.weight or default_weight,
-            }
-            -- docs: https://github.com/yidongnan/grpc-spring-boot-starter/pull/496
-            if is_grpc(scheme) and host.metadata and host.metadata.gRPC_port then
-                node.port = host.metadata.gRPC_port
+            local key = get_key(namespace_id, group_name, service_info.service_name)
+            service_names[key] = true
+
+            local hosts = data.hosts
+            if type(hosts) ~= 'table' then
+                hosts = {}
             end
 
-            core.table.insert(nodes, node)
-        end
-        if #nodes > 0 then
-            local content = core.json.encode(nodes)
-            nacos_dict:set(key, content)
+            local nodes = {}
+            for _, host in ipairs(hosts) do
+                local node = {
+                    host = host.ip,
+                    port = host.port,
+                    weight = host.weight or default_weight,
+                }
+                -- docs: https://github.com/yidongnan/grpc-spring-boot-starter/pull/496
+                if is_grpc(scheme) and host.metadata and host.metadata.gRPC_port then
+                    node.port = host.metadata.gRPC_port
+                end
+
+                core.table.insert(nodes, node)
+            end
+
+            if #nodes > 0 then
+                nodes_cache[key] = nodes
+            end
         end
-        ::CONTINUE::
     end
-    -- remove services that are not in use anymore
+
+    if not had_success then
+        return false, last_err or 'no available nacos services'
+    end
+
+    for key, nodes in pairs(nodes_cache) do
+        local content = core.json.encode(nodes)
+        nacos_dict:set(key, content)
+    end
+
     for key, _ in pairs(curr_service_in_use) do
         if not service_names[key] then
             nacos_dict:delete(key)
         end
     end
+
     curr_service_in_use = service_names
+    return true
+end
+
+
+local function fetch_full_registry(premature)
+    if premature then
+        return
+    end
+
+    local infos = get_nacos_services()
+    if #infos == 0 then
+        return
+    end
+
+    local host_list = local_conf.discovery.nacos.host
+    local host_count = #host_list
+    local start = math_random(host_count)
+    local last_err
+
+    for i = 0, host_count - 1 do
+        local idx = (start + i - 1) % host_count + 1
+        local base_uri, username, password = get_base_uri_by_index(idx)
+
+        if not base_uri then
+            last_err = 'invalid nacos host entry'
+            log.warn('nacos host at index ', idx, ' is invalid, skip')
+        else
+            local ok, err = fetch_from_host(base_uri, username, password, infos)
+            if ok then
+                return
+            end
+            last_err = err
+            log.warn('fetch_from_host failed from host ', base_uri, ': ', err)
+        end
+    end
+
+    log.error('failed to fetch nacos registry from all hosts: ', last_err)
```

Comment:
Every err will be printed, doesn't it seem necessary to print it again at the end?

Reply:
I removed `last_err` and now print each error in the loop, and if all errors occur, I print an error log at the end.